### PR TITLE
ExtremelyAsyn -restart tested

### PR DESCRIPTION
**Pull request by joannaqw**
1. Made a new class extremeAsynLASSCF, which is a subclass of LASSCFNoSymm
2. within the new class, modified rdm_cycle and ci_cycle to read RDMs, write Hamiltonians and break
3. added reading and writing orbitals (mo_coeff) within the kernel function

### CODE REVIEW

**Comment by MatthewRHermes** [comment]
Could you resolve conflicts please? (Pull master to your local branch, resolve the conflict as you see fit, commit, and push to your remote branch and it will update here.)

File: my_pyscf/mcscf/lasscf_rdm.py [hunk 1]
```diff
@@ -117,48 +118,60 @@ def get_init_guess_rdm (las, mo_coeff=None, h2eff_sub=None):
 def _ddm (dm1frs0, dm1frs1):
     return np.concatenate ([(d1 - d0).ravel () for d1, d0 in zip (dm1frs0, dm1frs1)])
 
-
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/lasscf_rdm.py:120]: What is the purpose of the changes to this function? It seems you replace it in your subclass anyway.

File: my_pyscf/mcscf/lasscf_rdm.py [hunk 2]
```diff
@@ -117,48 +118,60 @@ def get_init_guess_rdm (las, mo_coeff=None, h2eff_sub=None):
...
+            ddm = linalg.norm (ddm)
+            log.info ("LASSCF rdm-jk %d : E = %.15g ; dE = %.15g ; |ddm| = %.15g", it+1, e_tot,
+                      e_tot-e_old, ddm)
+            if (ddm < conv_tol_rdmjkddm) and (abs(e_tot-e_old) < conv_tol_rdmjkde):
+                converged = True
+                break
+        log.info ("LASSCF rdm-jk {}".format (('not converged','converged')[int(converged)]))
+        return e_cas, casdm1frs, casdm2fr, converged
 
 def kernel (las, mo_coeff=None, casdm1frs=None, casdm2fr=None, conv_tol_grad=1e-4, verbose=lib.logger.NOTE):
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/lasscf_rdm.py:160]: Editing this function changes the behavior of the existing lasscf_rdm implementation as well as your new implementation. Can you instead write a separate kernel function for your subclass, if you must change something in kernel?

Comment (MatthewRHermes) [reply my_pyscf/mcscf/lasscf_rdm.py:160]: Or, failing that, you could instead write "hook" functions that do nothing for the lasscf_rdm.LASSCFNoSymm class but which do the I/O for the lasscf_rdm.extremeAsynLASSCF class.

File: my_pyscf/mcscf/lasscf_rdm.py [hunk 3]
```diff
@@ -179,8 +192,8 @@ def kernel (las, mo_coeff=None, casdm1frs=None, casdm2fr=None, conv_tol_grad=1e-
     converged = False
     t2 = (t1[0], t1[1])
     it = 0
-    for it in range (las.max_cycle_macro):
-        e_cas, casdm1frs, casdm2fr, rdmjk_conv = rdm_cycle (las, mo_coeff, casdm1frs,
+    for it in range (las.max_cycle_macro): ###las.rdm_cycle
+        e_cas, casdm1frs, casdm2fr, rdmjk_conv = las.rdm_cycle (mo_coeff, casdm1frs,
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/lasscf_rdm.py:196]: This is kind of an example of a "hook" that my comment referred to above. `las.rdm_cycle` does one thing for lasscf_rdm.LASSCFNoSymm, and another thing for lasscf_rdm.extremeAsymLASSCF.

File: my_pyscf/mcscf/lasscf_rdm.py [hunk 4]
```diff
@@ -503,6 +525,116 @@ def kernel(self, mo_coeff=None, casdm1frs=None, casdm2fr=None, conv_tol_grad=Non
 
         return self.e_tot, self.e_cas, self.casdm1frs, self.casdm2fr, self.mo_coeff, self.mo_energy, h2eff_sub, veff
 
+class extremeAsynLASSCF (LASSCFNoSymm):
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/lasscf_rdm.py:528]: At a certain point, if this gets too big, it would make sense to put it in a new file.

File: my_pyscf/mcscf/lasci_sync.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 from scipy.sparse import linalg as sparse_linalg
 from scipy import linalg 
 import numpy as np
-
+import h5py
```
Comment (MatthewRHermes) [inline my_pyscf/mcscf/lasci_sync.py:7]: Is this necessary?

**Comment by joannaqw** [comment]
Going to keep this branch separated (on my fork) while Mario and I continue working on it.
